rbd/migration.go

```diff
@@ -0,0 +1,219 @@
+//go:build !(octopus || nautilus) && ceph_preview
+// +build !octopus,!nautilus,ceph_preview
+
+package rbd
+
+// #cgo LDFLAGS: -lrbd
+// #include <stdlib.h>
+// #include <rados/librados.h>
+// #include <rbd/librbd.h>
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/rados"
+)
+
+// MigrationImageState denotes the current migration status of a given image
+type MigrationImageState int
+
+const (
+	// MigrationImageUnknown is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_UNKNOWN from librbd.
+	MigrationImageUnknown = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_UNKNOWN)
+	// MigrationImageError is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ERROR from librbd.
+	MigrationImageError = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ERROR)
+	// MigrationImagePreparing is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARING from librbd.
+	MigrationImagePreparing = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARING)
+	// MigrationImagePrepared is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARED from librbd.
+	MigrationImagePrepared = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARED)
+	// MigrationImageExecuting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTING from librbd.
+	MigrationImageExecuting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTING)
+	// MigrationImageExecuted is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTED from librbd.
+	MigrationImageExecuted = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTED)
+	// MigrationImageAborting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ABORTING from librbd.
+	MigrationImageAborting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ABORTING)
+)
+
+// MigrationImageInfo provides information about the
+// live migration progress of an image
+type MigrationImageInfo struct {
+	SourcePoolID        int
+	SourcePoolNamespace string
+	SourceImageName     string
+	SourceImageID       string
+	DestPoolID          int
+	DestPoolNamespace   string
+	DestImageName       string
+	DestImageID         string
+	State               MigrationImageState
+	StateDescription    string
+}
+
+// MigrationPrepare prepares a migration
+// creating a target image with a link
+// to source and making source read-only
+//
+// Implements:
+//  int rbd_migration_prepare(rados_ioctx_t ioctx,
+//                            const char *image_name,
+// 	                          rados_ioctx_t dest_ioctx,
+//                            const char *dest_image_name,
+//                            rbd_image_options_t opts);
+//
+func MigrationPrepare(ioctx *rados.IOContext, destIoctx *rados.IOContext, sourceImageName string, destImageName string, rio *ImageOptions) error {
```

Comment:
Let's match that order of arguments in the c-call. So `ctx, name, ctx, name, ...` instead of `ctx, ctx, name, name`.  We usually try to stick to matching the underlying C apis for things like this.

Reply:
done